pkg/solana/chainreader/event_read_binding.go

```diff
@@ -0,0 +1,245 @@
+}
+
+func (b *eventReadBinding) SetCodec(codec types.RemoteCodec) {
+	b.codec = codec
+}
+
+func (b *eventReadBinding) SetModifier(modifier commoncodec.Modifier) {
+	b.modifier = modifier
+}
+
+func (b *eventReadBinding) CreateType(forEncoding bool) (any, error) {
+	itemType := codec.WrapItemType(forEncoding, b.namespace, b.genericName)
+
+	return b.codec.CreateType(itemType, forEncoding)
+}
+
+func (b *eventReadBinding) Decode(ctx context.Context, bts []byte, outVal any) error {
+	itemType := codec.WrapItemType(false, b.namespace, b.genericName)
+
+	return b.codec.Decode(ctx, bts, outVal, itemType)
+}
+
+func (b *eventReadBinding) QueryKey(
+	ctx context.Context,
+	filter query.KeyFilter,
+	limitAndSort query.LimitAndSort,
+	sequenceDataType any,
+) ([]types.Sequence, error) {
+	var (
+		pubKey solana.PublicKey
+		err    error
+	)
+
+	if pubKey, err = b.GetAddress(ctx, nil); err != nil {
+		return nil, err
+	}
+
+	if filter, err = b.remapper.remap(filter); err != nil {
+		return nil, err
+	}
+
+	// filter should always use the address and event sig
+	filter.Expressions = append([]query.Expression{
+		logpoller.NewAddressFilter(pubKey),
+		logpoller.NewEventSigFilter(b.eventSig[:]), // TODO: genericName is wrong; need the event sig hash
+	}, filter.Expressions...)
+
+	itemType := strings.Join([]string{b.namespace, b.genericName}, ".")
+
+	logs, err := b.events.FilteredLogs(ctx, filter.Expressions, limitAndSort, itemType)
+	if err != nil {
+		return nil, err
+	}
+
+	sequences, err := b.decodeLogsIntoSequences(ctx, logs, sequenceDataType)
+	if err != nil {
+		return nil, err
+	}
+
+	return sequences, nil
+}
+
+func (b *eventReadBinding) remapPrimitive(expression query.Expression) (query.Expression, error) {
+	var (
+		comp query.Expression
+		err  error
+	)
+
+	switch primitive := expression.Primitive.(type) {
+	case *primitives.Comparator:
+		if comp, err = b.encodeComparator(primitive); err != nil {
+			return query.Expression{}, fmt.Errorf("failed to encode comparator %q: %w", primitive.Name, err)
+		}
+	case *primitives.Confidence:
+		// TODO: maybe use an ignore filter?
+		// confidence is ignored in solana
+	}
+
+	return comp, nil
+}
+
+func (b *eventReadBinding) encodeComparator(comparator *primitives.Comparator) (query.Expression, error) {
+	onChainName, ok := b.indexedSubkeys[comparator.Name]
+	if !ok {
+		return query.Expression{}, fmt.Errorf("%w: unknown indexed subkey mapping %s", types.ErrInvalidConfig, comparator.Name)
+	}
+
+	itemType := strings.Join([]string{b.namespace, b.genericName, comparator.Name}, ".")
+
+	for idx, comp := range comparator.ValueComparators {
+		// need to do a transform and then extract the value for the subkey
+		newValue, err := b.modifier.TransformToOnChain(comp.Value, itemType)
+		if err != nil {
+			return query.Expression{}, err
+		}
+
+		comparator.ValueComparators[idx].Value = newValue
+	}
+
+	return logpoller.NewEventSubkeyFilter(strings.Split(onChainName, "."), comparator.ValueComparators), nil
```

Comment:
This is calling the wrong constrctor. Needs to call this constructor:
```suggestion
	return logpoller.NewEventBySubkeyFilter(subKeyIndex, comparator.ValueComparators), nil
```

We don't actually need to find the `onChainName` at all here, since the subKeyIndex is the only thing that goes in the actual query (it's what's used to index the subkey_values column, which is an array).

Unfortunately, it looks like we'll probably have to do a bit of refactoring here to find the index, since that needs to be looked up in the ChainReader config.  There should be an array of subkeys specified in the ChainReader config for this event.  What needs to happen is it finds the index in that array and passes that down to LogPoller. The path itself isn't as helpful, because in order to use it it would have to construct a more complex query that JOIN's the logs table with the filters table.

Reply:
I'm not following how we went from subkey paths to indexes.

Reply:
I left a bunch more comments, with code samples to illustrate what I'm suggesting and what I mean by the "subKeyIndex"

If you follow those suggestions or something like it, then all you'll need to pass here for the first arg to `NewEventBySubkeyFilter` is:
```
b.subKeyIndexes[comp.Value]
```